### ISSUE TITLE
ANR or CRASH when closing stream without complete stop

### ISSUE THREAD

**Issue by philburk** [google/oboe#227]
This report is actually from Diljeet from #214:
------------- BEGIN QUOTE
I am the developer of app named Harmonium
https://play.google.com/store/apps/details?id=harmonium.music.gameg.real.harmoniumfree
It has many users at least 100,000 current users with oboe

I want to help, but don't want the app to be listed on oboe page (not yet), once oboe is in stable version you can list my app on oboe page if you like.

I use oboe for one simple use only - play continuous frames like a piano app, so i only use simple output stream with low latency (notify me if you want to know all settings)
Most important - i don't set the Api, so it tries to use Aaudio on compatible devices.
I am writing this message so i can report the major problems i faced and how i solved them (with headaches and full day dry eye testings).

2 major problems faced on android 8.1 Aaudio devices

1) ANR- i used oboe start in activity onResume() (initialization and request start) and oboe stream ->close() in onPause()
This stream->close() call blocked and gave around 2000 ANRs in less than 7 days, i will provide log in end of message and solution too.

2) CRASH - after calling stream ->close() i had crashes in aaudio::AudioStreamLegacy::processCallbackCommon on specific devices but many and reported around 1900 crashes in less than 7 days, the problem - when user pressed home or exited the app with back the stream->close() was called and if the callback was still preparing the burst and returned after stream->close() is done i.e return DataCallbackResult::Continue; was called after stream->close() (not normal but happened) the app crashed with abort saying cxa_pure_virtual see log at bottom

1) ANR log
"main" prio=5 tid=1 Native
  | group="main" sCount=1 dsCount=0 flags=1 obj=0x727c1838 self=0x7329ac4e00
  | sysTid=6658 nice=-4 cgrp=default sched=0/0 handle=0x73afc489a8
  | state=S schedstat=( 23887114216 2007204581 38713 ) utm=1715 stm=673 core=0 HZ=100
  | stack=0x7fe4bf5000-0x7fe4bf7000 stackSize=8MB
  | held mutexes=
  #00  pc 000000000001d52c  /system/lib64/libc.so (syscall+28)
  #01  pc 0000000000067d54  /system/lib64/libc.so (pthread_cond_wait+96)
  #02  pc 000000000001168c  /system/lib64/libutils.so (android::Thread::requestExitAndWait()+84)
  #03  pc 00000000000433bc  /system/lib64/libaudioclient.so (android::AudioTrack::~AudioTrack()+428)
  #04  pc 00000000000437a8  /system/lib64/libaudioclient.so (android::AudioTrack::~AudioTrack()+16)
  #05  pc 000000000000d5f0  /system/lib64/libutils.so (android::RefBase::decStrong(void const*) const+112)
  #06  pc 000000000001cb10  /system/lib64/libaaudio.so (aaudio::AudioStreamTrack::~AudioStreamTrack()+88)
  #07  pc 000000000001b068  /system/lib64/libaaudio.so (AAudioStream_close+84)
  #08  pc 000000000004a7c8  /data/app/harmonium.music.gameg.real.harmoniumfree-FiHTwb0Ip2FBG_5ZmvGfDw==/lib/arm64/libndk1.so (oboe::AudioStreamAAudio::close()+64)
  #09  pc 000000000004409c  /data/app/harmonium.music.gameg.real.harmoniumfree-FiHTwb0Ip2FBG_5ZmvGfDw==/lib/arm64/libndk1.so (MyOboeEngine::stop()+32)
  #10  pc 000000000000084c  /data/app/harmonium.music.gameg.real.harmoniumfree-FiHTwb0Ip2FBG_5ZmvGfDw==/oat/arm64/base.odex (Java_harmonium_music_gameg_real_harmoniumfree_HarmoniumActivity_stopEngine__+124)
  at harmonium.music.gameg.real.harmoniumfree.HarmoniumActivity.stopEngine (HarmoniumActivity.java)
  at harmonium.music.gameg.real.harmoniumfree.HarmoniumActivity.onPause (HarmoniumActivity.java:7284)
  at android.app.Activity.performPause (Activity.java:7286)

There are many more ANRs similar to this, i.e. the all started with stream->close()

2) CRASH log
Today, 1:57 AM on app version 17
Xiaomi Redmi 5 Plus (vince), Android 8.1
Report 1

*** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
pid: 0, tid: 0 >>> harmonium.music.gameg.real.harmoniumfree <<<

backtrace:
  #00  pc 000000000001de54  /system/lib64/libc.so (abort+104)
  #01  pc 0000000000043408  /system/lib64/libc++.so (abort_message+244)
  #02  pc 000000000006bb40  /system/lib64/libc++.so (__cxa_pure_virtual+16)
  #03  pc 000000000001bae8  /system/lib64/libaaudio.so (aaudio::AudioStreamLegacy::processCallbackCommon(int, void*)+352)
  #04  pc 00000000000474d0  /system/lib64/libaudioclient.so (android::AudioTrack::processAudioBuffer()+1768)
  #05  pc 00000000000492a0  /system/lib64/libaudioclient.so (android::AudioTrack::AudioTrackThread::threadLoop()+192)
  #06  pc 0000000000011538  /system/lib64/libutils.so (android::Thread::_threadLoop(void*)+280)
  #07  pc 0000000000067e50  /system/lib64/libc.so (__pthread_start(void*)+36)
  #08  pc 000000000001f280  /system/lib64/libc.so (__start_thread+68)

All the 1900 crahes has exact same log only sometimes the +352 changed maybe because of version updates

Solutions Now or i should say hacks
1) ANR solution - simple but tricky, i somehow managed to get the stream->close() to be called on a seperate thread and the ANR is gone now
2) CRASH solution : i should say it is simple but i had a very hard time figuring this out, had to borrow a device too
	the stopPhase2() is the solution i.e. calling stop before close, it waits until the audio call back is processing and then closes the stream
Detail:
- on Activity pause - 
java code in onPause()

        boolean phase2Required = stopPhase1();
        if(phase2Required){new Thread(){
            @Override
            public void run() {
                stopPhase2(); // actually called through jni from another jni function
            }
        }.start();}

c++ code note stopPhase2() is called on a thread

	AudioStream *mStream = nullptr, *mStreamToDispose = nullptr;

	bool MyOboeEngine::stopPhase1() {
		AudioStream *mStreamCopy = mStream;
		mStream = nullptr; // reason for setting null quickly check onAudioReady
		bool phase2Required = false;

		if (mStreamCopy != nullptr){
			phase2Required = true;
			if(mStreamToDispose == nullptr) {
				mStreamToDispose = mStreamCopy;
				mStreamCopy->requestStop(); // for non blocking operation
			}else{
				//try normal approach
				mStreamCopy->stop();
				mStreamCopy->close();
			}
			//__android_log_print(ANDROID_LOG_ERROR,"oboelog:","Stream Closed - Phase 1");
		}
		return phase2Required;
	}
	void MyOboeEngine::stopPhase2() {
		AudioStream *mStreamToDisposeCopy = mStreamToDispose;
		mStreamToDispose = nullptr;
		if (mStreamToDisposeCopy != nullptr){
			if(mStreamToDisposeCopy->getState() != StreamState::Stopped){
				mStreamToDisposeCopy->stop();
				double msMax=currentTimeMillisDouble()+25;
				//wait for all callbacks to finish in 25 ms
				if(msMax>30)
				while (callBackCallCount!=0 && currentTimeMillisDouble()<msMax){}
			}
			mStreamToDisposeCopy->close();
			//__android_log_print(ANDROID_LOG_ERROR,"oboelog:","Stream Closed - Phase 2");
		}
	}
	DataCallbackResult
	MyOboeEngine::onAudioReady(AudioStream *oboeStream, void *audioData, int32_t numFrames) {
		callBackCallCount++;
		if(mStream == nullptr) return DataCallbackResult::Stop;

		...... audio processing code

		callBackCallCount--;
		return mStream == nullptr ? DataCallbackResult::Stop : DataCallbackResult::Continue;
	  // important is return Stop if mStream == nullptr i check it in start
	}

i hope this code is clear and easy, please paste it in editor and format it for clearity

Also please correct me and guide me for a better implementation instead of hacks

_Originally posted by @Diljeet in https://github.com/google/oboe/issues/214#issuecomment-423704959_

**Comment by Diljeet** [comment]
One more detail about the "2. CRASH" - i was able to generate the crash (cxa_pure_virtual pure virtual method called) on my device in 2 ways.

i initialize engine in onResume() and stream->close in onPause()
1) i pressed home and launched again my app, quickly several times, home and launch app again and again, and i had the crash logged in above message on my device. Only 2 to 3 times it happened in a long while but when i did it on the borrowed Redmi S2 device it happened in every 4 to 5 times.
2) Not Genuine: i was trying to manually generate the crash by making errors intentionally and i could generate the crash by this method, its a wrong code but it generated the crash for me.

```
DataCallbackResult
MyOboeEngine::onAudioReady(AudioStream *oboeStream, void *audioData, int32_t numFrames) {
	
	oboeStream->close();// after calling this the crash occured
	
	return DataCallbackResult::Continue;  
}
```

this clarified that the crash was happening if the stream->close() was called before the function returned, this method is just to see the crash not an actual case in the app. Good Luck in resolving the issues.

About code i almost followed the examples in the oboe project

List of devices the crash majorly occurred.
```
abort crash on  android 8.1
vivo 1724 (1724)	120	8.0%
Galaxy J7 Neo (j7velte)	110	7.4%
Y71 (1801)	97	6.5%
CAMON iSky (TECNO-IN2)	93	6.2%
CPH1803 (CPH1803)	89	6.0%
Redmi S2 (ysl)	79	5.3%
Redmi Note 5 Pro (whyred)	69	4.6%
Z61 (Z61)	66	4.4%
vivo 1802 (1802)	60	4.0%
Mi A1 (tissot_sprout)	54	3.6%
Mi A2
Z50 (Z50)	53	3.5%
Redmi 5 Plus (vince)	49	3.3%
vivo 1803 (1803)	39	2.6%
CAMON I Ace (TECNO-IN1)	32	2.1%
vivo 1723 (1723)	30	2.0%
vivo 1726 (1726)	25	1.7%
vivo 1727 (1727)	25	1.7%
CPH1819 (CPH1819)	25	1.7%
RMX1805 (RMX1805)	24	1.6%
HUAWEI Y5 Prime 2018 (HWDRA-M)	20	1.3%
```

There were other devices too, if anyone wants then i can check their device name in the logs.
Good Day. Best of Luck to all.

**Comment by Diljeet** [comment]
This particular crash happened in android 8.1 only (pure virtual method called), but according to other crashes - similar situations also happened on other Android versions 6.0, 7.0 in reports but these crashes had very low count and had different logs (definitely they happened due to callback still going on i.e. in processCallback).

About calling close() in callback - it was one of the many intentionally generated errors i was creating just to get the exact crash on my device. Never used it in my app.

And as for android 8.0 on my device custom rom (from xda for testing), oboe chooses to use OpenSL ES instead of Aaudio, maybe it does the same on other phones too. This maybe the reason i get Aaudio crashes only 8.1.

Lastly 9.0 is probably new so no crashes now. I did checked my app on Android 9.0 custom rom and it worked great with no crashes and very low latency, with and without headphones. Thanks for your support, Sir.

**Comment by philburk** [comment]
I cannot reproduce this error.

I added a 500 msec sleep in my callback and then timed the close() so that it occurred in the middle of the callback. When AAudio closed the track it stopped the stream an then waited for my slow callback to finish. That is the requestExitAndWait() call.

    void close() {
        LOGI("%s() set sleepUSec", __func__);
        if (oboeCallbackProxy != nullptr) {
            oboeCallbackProxy->sleepUSec.store(500 * 1000); // HACK - request delay
            usleep(200 * 1000); // sleep until we know we are in the middle of the callback
        }
        if (oboeStream != nullptr) {
            LOGI("%s() close oboeStream", __func__);
            oboeStream->close();
            LOGI("%s() finished closing oboeStream", __func__);
        }
        delete oboeStream;
        oboeStream = nullptr;
        delete oboeCallbackProxy;
        oboeCallbackProxy = nullptr;

The log ---------------

09-26 08:11:47.518  7088  7088 I OboeTester: close() set sleepUSec
09-26 08:11:47.527  7088  7152 I OboeAudio: onAudioReady() callback sleeping
09-26 08:11:47.719  7088  7088 I OboeTester: close() close oboeStream
09-26 08:11:47.719  7088  7088 D AAudio  : AAudioStream_close(0x7c87884f00)
09-26 08:11:47.719  7088  7088 D AudioTrack: stop() called with 448292 frames delivered
09-26 08:11:47.804   646  1178 E volume_listener: check_and_set_gain_dep_cal: Failed to set gain dep cal level
09-26 08:11:47.805   646  1103 E volume_listener: check_and_set_gain_dep_cal: Failed to set gain dep cal level

AudioTrack close seems to block here until the callback returns.

09-26 08:11:48.027  7088  7152 I OboeAudio: onAudioReady() callback woke up

Ran a second time! I think so it can acknowledge the exit request.

09-26 08:11:48.028  7088  7152 I OboeAudio: onAudioReady() callback sleeping
09-26 08:11:48.528  7088  7152 I OboeAudio: onAudioReady() callback woke up
09-26 08:11:48.529  7088  7088 D AAudioStream: destroying 0x7c87884f00, state = AAUDIO_STREAM_STATE_CLOSED
09-26 08:11:48.529  7088  7088 I OboeTester: close() finished closing oboeStream
09-26 08:11:48.529  7088  7088 I OboeTester: close() finished deleting everything

I increased the sleep to 3000 msec and it still worked.
I tested with MMAP and it also waited for the callback to finish by joining the thread.

**Comment by philburk** [comment]
@Diljeet - can you tell me more about your app?

The ANR suggests that your callback may be taking a very long time or hanging. 
Is there something that happens during the close that could call the callback to wait or sleep or spin or hang?

I am concerned about the way that your onAudioReady() checks (mStream == nullptr).
If it uses mStream anywhere else in the callback then there could be a race condition if close is called when the callback is running. I think it would be best to only set mStream to nullptr after the close has finished.

You do not need to tell your callback to return DataCallbackResult::Stop. That is only for when the callback decides to stop. An example might be when the callback has finished playing a particular sound.

> i used oboe start in activity onResume() (initialization and request start) 
> and oboe stream ->close() in onPause()

These are not balanced. Maybe try this for the Activity Life Cycle:

onStart() - open oboeStream
onResume() - oboeStream->requestStart()

onPause() - oboeStream->requestStop()
onStop() - oboeStream->close()

**Comment by Diljeet** [comment]
Hi i am sorry for late reply, i didn't had any email notification about your reply. I just accidentally checked the issue and found your replies, so now i will try to give all possible information.

You are right sir, i also can't produce the error by sleep or waiting on my device, it only happened randomly on the above mentioned devices. I mentioned earlier it could happen when you repeatedly press home and launch app again (on some devices only).

About my app, its named Harmonium in the Android play store, i will explain the coding now - i have pre-calculated arrays, which play repeatedly when someone presses a key, few calculations are done before playing the arrays, 1) a simple 4 point fast cubic resampler and 2) a equalizer if someone enabled it, which rarely happens. all the calculations takes less than 0.25 of 0.5 milli second averagely for a 5 ms burst.

* i think i am not doing any hanging tasks in onAudioReady as its a synthesizer app and i removed all unnecessary calculations from the onAudioReady callback.

i use mono output
	
```
	DataCallbackResult
	MyOboeEngine::onAudioReady(AudioStream *oboeStream, void *audioData, int32_t numFrames) {
		if(mStream == nullptr) return DataCallbackResult::Stop;
		
		//clear arrays
		clearArrays();
		
		//resample and set the array from precalculated arrays very fast...
		setPlaybackData();
		
		if(keyCount > 0) {
			//apply equalizer if necessary
			if (equalizer==1)
				applyEqualizer();
				
			//check and normalizer if clipping 
			checkArray();
		
			if(oboeStream->getFormat() == AudioFormat::Float){
				float *buffer = static_cast<float *>(audioData);
				convert & set array to float function
			} else {
				short *buffer = static_cast<short *>(audioData);
				convert & set array to short function
			}       
		}else{
			if(oboeStream->getFormat() == AudioFormat::Float){
				float *buffer = static_cast<float *>(audioData);
				for (int i = 0; i < numFrames; i++)
					buffer[i] = 0;
			} else {
				short *buffer = static_cast<short *>(audioData);
				for (int i = 0; i < numFrames; i++)
					buffer[i] = 0;
			}
		}
		return mStream == nullptr ? DataCallbackResult::Stop : DataCallbackResult::Continue;
	}
```

about setting it to null pointer, i never the reference of mStream in the callback, i even removed the oboeStream->getFormat()
and then i backup the mStream pointer before setting it to null and then close it.

about the onStart and onStop
i wlll surely change the initialization process from onResume to onStart, and onPause to onStop for avoiding a few cases when the activity is only paused and not stopped.
Good Day Sir, sorry for causing you trouble and delay.

**Comment by philburk** [comment]
It looks like this is resolved. If you see the problem again then please open a new issue with simple steps to reproduce.

**Comment by binodrok** [comment]
Hi, I am facing similar problem, Its not at all crashing but giving ANR while closing the oboe::AudioStream.

ANR  I am getting is,

`main" prio=5 tid=1 Native
  | group="main" sCount=1 dsCount=0 flags=1 obj=0x72abc5d8 self=0x729ef87000
  | sysTid=15040 nice=-10 cgrp=default sched=0/0 handle=0x72a04faee8
  | state=S schedstat=( 9723783801 2542767582 22171 ) utm=678 stm=294 core=2 HZ=100
  | stack=0x7ff793a000-0x7ff793c000 stackSize=8192KB
  | held mutexes=
  #00  pc 000000000008033c  /apex/com.android.runtime/lib64/bionic/libc.so (syscall+28)
  #01  pc 0000000000083e04  /apex/com.android.runtime/lib64/bionic/libc.so (__futex_wait_ex(void volatile*, bool, int, bool, timespec const*)+140)
  #02  pc 00000000000e6454  /apex/com.android.runtime/lib64/bionic/libc.so (pthread_cond_wait+60)
  #03  pc 0000000000013818  /system/lib64/libutils.so (android::Thread::requestExitAndWait()+84)
  #04  pc 000000000007880c  /system/lib64/libaudioclient.so (android::AudioTrack::~AudioTrack()+1252)
  #05  pc 00000000000780f4  /system/lib64/libaudioclient.so (android::AudioTrack::~AudioTrack()+16)
  #06  pc 000000000000f6b0  /system/lib64/libutils.so (android::RefBase::decStrong(void const*) const+112)
  #07  pc 000000000002bb34  /system/lib64/libaaudio.so (aaudio::AudioStreamTrack::~AudioStreamTrack()+184)
  #08  pc 0000000000028990  /system/lib64/libaaudio.so (aaudio::AudioStreamTrack::~AudioStreamTrack()+16)
  #09  pc 000000000002164c  /system/lib64/libaaudio.so (AAudioStream_close.cfi+452)
  #10  pc 0000000000075308  /data/app/****** (oboe::AudioStreamAAudio::close()+72)`

I checked all the possible races condition but there is not any races condition in my app.
While oboe::AudioStream ->close() is waiting I can see 
'D/OboeAudio: AudioStreamBuffered::onDefaultCallback(): xrun! framesTransferred = 0, numFrames = 192'
back to back in the log.
Which means callback is still getting called.

Before close I tried following
'   oboe::Result result = oboe::Result::OK;
    oboe::StreamState inputState = oboe::StreamState::Stopping;
    oboe::StreamState nextState = oboe::StreamState::Closing;
    int64_t timeoutNanos = 5000*kNano;
    if (mPlayStream) {
        result = mPlayStream->requestStop();
        result = mPlayStream->waitForStateChange(inputState, &nextState, timeoutNanos);
        result = mPlayStream->close();
}'

also
'       result = mPlayStream->stop(5*kNano);
        result = mPlayStream->close();
'

Can someone help me to resolve this issue?

**Comment by romanbsd** [comment]
Just observed this crash on my Xiaomi A3 (Android One device). Putting `stop` before `close` seems to fix it:
```
    --------- beginning of crash
2020-10-01 10:48:02.773 6907-7263/com.piano.keyboard A/libc: Fatal signal 6 (SIGABRT), code -6 (SI_TKILL) in tid 7263 (AudioTrack), pid 6907 (.piano.keyboard)
2020-10-01 10:48:03.182 7370-7370/? I/crash_dump64: obtaining output fd from tombstoned, type: kDebuggerdTombstone
2020-10-01 10:48:03.196 1057-1057/? I//system/bin/tombstoned: received crash request for pid 7263
2020-10-01 10:48:03.200 7370-7370/? I/crash_dump64: performing dump of process 6907 (target tid = 7263)
2020-10-01 10:48:03.244 7370-7370/? A/DEBUG: *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
2020-10-01 10:48:03.245 7370-7370/? A/DEBUG: Build fingerprint: 'Xiaomi/laurel_sprout/laurel_sprout:10/QKQ1.190910.002/V11.0.21.0.QFQMIXM:user/release-keys'
2020-10-01 10:48:03.246 7370-7370/? A/DEBUG: Revision: '0'
2020-10-01 10:48:03.246 7370-7370/? A/DEBUG: ABI: 'arm64'
2020-10-01 10:48:03.268 7370-7370/? A/DEBUG: Timestamp: 2020-10-01 10:48:03+0300
2020-10-01 10:48:03.269 7370-7370/? A/DEBUG: pid: 6907, tid: 7263, name: AudioTrack  >>> com.piano.keyboard <<<
2020-10-01 10:48:03.269 7370-7370/? A/DEBUG: uid: 10324
2020-10-01 10:48:03.269 7370-7370/? A/DEBUG: signal 6 (SIGABRT), code -6 (SI_TKILL), fault addr --------
2020-10-01 10:48:03.269 7370-7370/? A/DEBUG: Abort message: 'Pure virtual function called!'
2020-10-01 10:48:03.269 7370-7370/? A/DEBUG:     x0  0000000000000000  x1  0000000000001c5f  x2  0000000000000006  x3  00000072908bc820
2020-10-01 10:48:03.269 7370-7370/? A/DEBUG:     x4  fefeff718f8abf57  x5  fefeff718f8abf57  x6  fefeff718f8abf57  x7  7f7f7f7fffffff7f
2020-10-01 10:48:03.269 7370-7370/? A/DEBUG:     x8  00000000000000f0  x9  30c67b86fab42644  x10 0000000000000001  x11 0000000000000000
2020-10-01 10:48:03.269 7370-7370/? A/DEBUG:     x12 fffffff0fffffbdf  x13 2164656c6c616320  x14 0000000000000004  x15 0000000000000000
2020-10-01 10:48:03.269 7370-7370/? A/DEBUG:     x16 0000007324d218c0  x17 0000007324cfd900  x18 0000007214d8a000  x19 0000000000001afb
2020-10-01 10:48:03.269 7370-7370/? A/DEBUG:     x20 0000000000001c5f  x21 00000000ffffffff  x22 00000072908bca70  x23 0000007324d1b620
2020-10-01 10:48:03.269 7370-7370/? A/DEBUG:     x24 00000072908bc950  x25 00000072908bc990  x26 000000000000aef0  x27 00000000000000c0
2020-10-01 10:48:03.269 7370-7370/? A/DEBUG:     x28 0000000000000009  x29 00000072908bc8c0
2020-10-01 10:48:03.269 7370-7370/? A/DEBUG:     sp  00000072908bc800  lr  0000007324caf0c4  pc  0000007324caf0f0
2020-10-01 10:48:03.322 7370-7370/? A/DEBUG: backtrace:
2020-10-01 10:48:03.322 7370-7370/? A/DEBUG:       #00 pc 00000000000830f0  /apex/com.android.runtime/lib64/bionic/libc.so (abort+160) (BuildId: cf739dbc84bcc78f7a1500721bfb3758)
2020-10-01 10:48:03.322 7370-7370/? A/DEBUG:       #01 pc 00000000000500fc  /system/lib64/libc++.so (abort_message+232) (BuildId: 5cead07f137a1a3575630376c60256c8)
2020-10-01 10:48:03.322 7370-7370/? A/DEBUG:       #02 pc 0000000000067204  /system/lib64/libc++.so (__cxa_pure_virtual+16) (BuildId: 5cead07f137a1a3575630376c60256c8)
2020-10-01 10:48:03.323 7370-7370/? A/DEBUG:       #03 pc 0000000000024ef4  /system/lib64/libaaudio.so (aaudio::AudioStreamLegacy::processCallbackCommon(int, void*)+1168) (BuildId: 810627d5d7cacc093df16f9834e25092)
2020-10-01 10:48:03.323 7370-7370/? A/DEBUG:       #04 pc 000000000007af78  /system/lib64/libaudioclient.so (android::AudioTrack::processAudioBuffer()+3044) (BuildId: 45a2142bc541f45ae2e6af7c1d82b52d)
2020-10-01 10:48:03.323 7370-7370/? A/DEBUG:       #05 pc 000000000007a164  /system/lib64/libaudioclient.so (android::AudioTrack::AudioTrackThread::threadLoop()+484) (BuildId: 45a2142bc541f45ae2e6af7c1d82b52d)
2020-10-01 10:48:03.323 7370-7370/? A/DEBUG:       #06 pc 0000000000013588  /system/lib64/libutils.so (android::Thread::_threadLoop(void*)+224) (BuildId: 8dcc6a23d7b5c85e8562aa5394a3a42f)
2020-10-01 10:48:03.323 7370-7370/? A/DEBUG:       #07 pc 00000000000c2b8c  /system/lib64/libandroid_runtime.so (android::AndroidRuntime::javaThreadShell(void*)+140) (BuildId: 4c576daabd24c4da875c383c2f63f3fb)
2020-10-01 10:48:03.323 7370-7370/? A/DEBUG:       #08 pc 00000000000e6890  /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+36) (BuildId: cf739dbc84bcc78f7a1500721bfb3758)
2020-10-01 10:48:03.323 7370-7370/? A/DEBUG:       #09 pc 0000000000084b6c  /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+64) (BuildId: cf739dbc84bcc78f7a1500721bfb3758)
```

**Comment by romanbsd** [comment]
Is it related to https://github.com/google/oboe/issues/961 ?
